[MRG] Backports msg in assert_raises and assert_raises_regex

thechargedneutron opened the pull request:
#### Reference Issue
Fixes #9454 

#### What does this implement/fix? Explain your changes.
I added a new directory named `modifiedunittest` which backports `msg`.

#### Any other comments?
1. I have not removed the extra codes of the pasted module, will be done upon validation of my task.
2. This should work with Python 2.6 as well. If it fails, necessary improvement will be done.
Thanks

thechargedneutron committed:
Added modifiedunittest

thechargedneutron committed:
Backports msg in assertRaises and assertRaisesRegexp

thechargedneutron committed:
Import statement corrected

thechargedneutron committed:
Corrected import statement

thechargedneutron commented:
After having a look at the error log, which states `ImportError`, what all changes do I need to make in order to recognise the package I used. 
One thing I observed was to add  `config.add_subpackage('modifiedunittest')` in `sklearn.utils.setup.py` . Any other location where I must add something?

thechargedneutron committed:
Added module name in utils.setup.py

thechargedneutron committed:
Merge branch 'Secondary' of https://github.com/thechargedneutron/scikit-learn into Secondary

jnothman commented on sklearn/utils/modifiedunittest/case.py:
```diff
@@ -0,0 +1,1176 @@
+    return isinstance(expected, type) and issubclass(expected, basetype)
+
+
+class _AssertRaisesContext(_AssertRaisesBaseContext):
```
I think we basically just want this (perhaps with some changes for compatibility), and nothing else.

thechargedneutron replied:
Okay, I'll remove all the unnecessary packages once I am sure the whole works. I am getting ImportError on writing `import sklearn.utils.modifiedunittest` . Please help me in importing it correctly. I'll remove the extra code here.

thechargedneutron commented:
So, now I realise that we don't need to import the modules and all. We can just copy paste the codes in `testing.py` itself since you pointed out that there's only a few things to be added. Is this a good idea or importing from other file is better.

thechargedneutron committed:
Removed Extra modules

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into Secondary

thechargedneutron committed:
Reordered class

thechargedneutron committed:
_is_subtype added

thechargedneutron committed:
Missing import added

thechargedneutron committed:
_formatMessage added

thechargedneutron committed:
missing variables added

thechargedneutron committed:
Remove PEP8 failures

thechargedneutron committed:
Removed safe_repr

thechargedneutron commented:
I guess now we have the minimum amount of code in `sklearn.utils.testing.py`. The extra module added is removed. Hope this works fine.
Sorry for so many commits.

jnothman commented:
many commits is not a problem. I'll try look soon

On 14 Aug 2017 6:59 am, "Kumar Ashutosh" <notifications@github.com> wrote:

> I guess now we have the minimum amount of code in sklearn.utils.testing.py.
> The extra module added is removed. Hope this works fine.
> Sorry for so many commits.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9536#issuecomment-322067161>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz64ahlMLxhQ7e0s6KoPcW_idl4Kxjks5sX2O-gaJpZM4O1YVM>
> .
>

thechargedneutron commented:
@jnothman  Does it work or there's something more to be done here? 
The codecov patch test fails. But I don't think we need the test function of the functions added by me. I need your opinion in this.

amueller commented on sklearn/utils/testing.py:
```diff
@@ -68,12 +69,184 @@
            "assert_approx_equal", "SkipTest"]
 
 
+def _is_subtype(expected, basetype):
```
Maybe it would be better to put this in a separate file like ``_unittest_backport.py`` or something?

thechargedneutron replied:
Okay, I'll make a new file for this.

amueller commented:
I think we should have a small test for this.

thechargedneutron commented:
Is the backport implementation working? I wanted to know this. :)
I will add tests as soon as possible.

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into Secondary

thechargedneutron committed:
_unittest_backport.py added

thechargedneutron committed:
Import statement corrected

thechargedneutron commented:
I think I need some help in making the test functions. 
In python's source code, I found the following code: 
Is this the appropriate test case for `_formatMessage` function which is there in `class TestCase_new` ?
Kindly help as I am not sure about the test functions.

``` 
class TestLongMessage(unittest.TestCase):
    def setUp(self):
        class TestableTestFalse(unittest.TestCase):
            longMessage = False
            failureException = self.failureException

            def testTest(self):
                pass

        class TestableTestTrue(unittest.TestCase):
            longMessage = True
            failureException = self.failureException

            def testTest(self):
                pass

        self.testableTrue = TestableTestTrue('testTest')
        self.testableFalse = TestableTestFalse('testTest')

    def test_formatMsg(self):
        self.assertEqual(self.testableFalse._formatMessage(None, "foo"), "foo")
        self.assertEqual(self.testableFalse._formatMessage("foo", "bar"), "foo")

        self.assertEqual(self.testableTrue._formatMessage(None, "foo"), "foo")
        self.assertEqual(self.testableTrue._formatMessage("foo", "bar"), "bar : foo")

        # This blows up if _formatMessage uses string concatenation
        self.testableTrue._formatMessage(object(), 'foo')
```

jnothman commented on sklearn/utils/_unittest_backport.py:
```diff
@@ -0,0 +1,173 @@
+import re
```
This should have attribution to where it is backported from (and the copyright notice)

jnothman commented on sklearn/utils/_unittest_backport.py:
```diff
@@ -0,0 +1,173 @@
+        return True
+
+
+class TestCase_new(object):
```
If you inherit from `unittest.TestCase` you can use this alone as the dummy, I think

thechargedneutron replied:
Something like this? `class TestCase_new(unittest.TestCase):`. And then we can remove the redundant functions?

thechargedneutron replied:
This somehow gives error. It says `ValueError: no such test method in <class 'TestCase_new'>: runTest` . I looked for the error in StackOverflow and found this: https://stackoverflow.com/questions/2090479/valueerror-no-such-test-method-in-class-myapp-tests-sessiontestcase-runtes   but could not make use of it. Can you suggest me how to achieve this?

jnothman reviewed (commented):
Thanks.

I don't think TestLongMessage is testing assertRaises.

jnothman commented:
to test you want something like:

```
def test_assert_raises_msg():
    with assert_raises_regex(AssertionError, 'Hello world'):
        with assert_raises(ValueError, msg='Hello world'):
            # Nothing is raised. Assertion should fail.
            pass
```

thechargedneutron commented:
@jnothman  Okay, so I was under the impression that I have to add tests for the functions in `_unittest_backport.py` and not for `assert_raises` . If only I add tests for assert_raises_with_msg, the codecov/patch will still fail, isn't it?

jnothman commented:
Yes, you won't get good codecov coverage. That's okay for a backport.

On 16 August 2017 at 12:59, Kumar Ashutosh <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> Okay, so I was under the
> impression that I have to add tests for the functions in
> _unittest_backport.py and not for assert_raises . If only I add tests for
> assert_raises_with_msg, the codecov/patch will still fail, isn't it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9536#issuecomment-322652854>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61VbBbmWBS2njUIXsHYQlCHgcaeDks5sYlssgaJpZM4O1YVM>
> .
>

thechargedneutron committed:
Added copyright

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into Secondary

thechargedneutron committed:
Syntax Error removed

thechargedneutron committed:
Error removed

jnothman commented on sklearn/utils/_unittest_backport.py:
```diff
@@ -0,0 +1,224 @@
+        return True
+
+
+class TestCase_new(unittest.TestCase):
```
you could just call the new class TestCase. Certainly, it shouldn't have an underscore in a class name.

thechargedneutron replied:
Okay, I'll do that. somehow I am not able to use a single dummy variable, initializing unittest.Test case('__init__') inside new TestCase gives an error saying the '__init__' function take 2 arguments, 4 given. Please look into it

thechargedneutron committed:
runTest function added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into Secondary

thechargedneutron committed:
Tests added

thechargedneutron committed:
__init__ added

thechargedneutron committed:
Import added

thechargedneutron commented:
So, I have made all the changes suggested so far, added `test_assert_raises_with_msg` , added copyright in `_unittest_backport.py` and also imported unittest module in my backported class. 
Kindly review the changes and suggest if anything else is required.

amueller commented:
LGTM

jnothman commented:
LGTM, thanks!

thechargedneutron commented:
Thanks a lot, this was a very good learning experience, will continue to contribute :)

jnothman commented:
Glad to hear it!

On 17 August 2017 at 19:25, Kumar Ashutosh <notifications@github.com> wrote:

> Thanks a lot, this was a very good learning experience, will continue to
> contribute :)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9536#issuecomment-323017229>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_m4L9GuyFfLeWyV8D_D8kJiS03Zks5sZAcYgaJpZM4O1YVM>
> .
>
